File: 1164-missing-number-in-arithmetic-progression/missing-number-in-arithmetic-progression.java
/*

Linear Search : O(n)
Binary Search: O(log(n))

*/


class Solution {
    public int missingNumber(int[] arr) {
        int n = arr.length;
        int diff = (arr[n-1] - arr[0]) / n;
        int start = 0, end = n - 1;
        while(start < end) {
            int mid = start + (end - start)/2;
            if(arr[mid] == arr[0] + (mid * diff)) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return arr[0] + (end * diff);
    }
}

File: 1164-missing-number-in-arithmetic-progression/missing-number-in-arithmetic-progression.py
class Solution:
    def missingNumber(self, arr: List[int]) -> int:
        n = len(arr)
        difference = int((arr[n-1] - arr[0]) / n)
        start, end = 0, n-1
        while start < end:
            mid = int(start + (end - start) / 2);
            if arr[mid] == (arr[0] + (mid * difference)):
                start = mid + 1
            else:
                end = mid;
        return arr[0] + (start * difference)